File: runtime/py/nightseam/runtime/identity.py
"""Declaration agreement as an ordinary, transport-independent request."""

import re

from .json import scalar_value
from .peer import PublicError

IDENTITY_METHOD = "identity.check"


def _read_identity(raw):
    def invalid(message):
        raise PublicError("contract_invalid", message)

    if not isinstance(raw, dict):
        invalid("a declaration identity is an object with path and optional digest")
    for key in raw:
        if key not in ("path", "digest"):
            invalid("identity: unknown field " + str(key))
    path = raw.get("path")
    if not isinstance(path, str) or not path:
        invalid("an identity names a nonempty Unicode declaration path")
    try:
        scalar_value(path)
    except ValueError:
        invalid("an identity names a nonempty Unicode declaration path")
    result = {"path": path}
    if "digest" in raw:
        digest = raw["digest"]
        if not isinstance(digest, str) or not re.fullmatch("[0-9a-f]{64}", digest):
            invalid("a declaration digest is lowercase SHA-256 hex")
        result["digest"] = digest
    return result


def _compare(local, remote):
    if local["path"] != remote["path"] or (
        "digest" in local and "digest" in remote and local["digest"] != remote["digest"]
    ):
        raise PublicError("contract_mismatch", "the declaration identity for " + local["path"] + " differs")


def identity_handler(expected):
    """Snapshot an identity and answer requests without running model code."""
    local = _read_identity(expected)

    def handle(value, context=None):
        _compare(local, _read_identity(value))
        return dict(local)

    return handle


async def check_identity(call, expected, *, timeout_ms=30_000, context=None):
    """Check before exposing a model; only method_not_found means absence.

    The supplied call must honor its timeout and cancellation. Refusal leaves
    the carrier open, and unspecified digests do not compare revisions.
    """
    local = _read_identity(expected)
    try:
        remote = await call(IDENTITY_METHOD, local, timeout_ms=timeout_ms, context=context)
    except PublicError as error:
        if error.code == "method_not_found":
            return
        raise
    _compare(local, _read_identity(remote))

File: runtime/hs/src/Nightseam/Runtime/Identity.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE ScopedTypeVariables #-}
module Nightseam.Runtime.Identity (DeclarationIdentity(..), identityHandler, checkIdentity) where
import Control.Exception hiding (Handler)
import Control.Monad
import Data.Aeson hiding (Options, defaultOptions)
import qualified Data.Aeson.KeyMap as KM
import Data.Text (Text)
import qualified Data.Text as T
import Nightseam.Runtime

data DeclarationIdentity = DeclarationIdentity { declarationPath :: Text, declarationDigest :: Text } deriving (Eq,Show)
instance ToJSON DeclarationIdentity where
  toJSON (DeclarationIdentity path digest) = object (["path" .= path] ++ if T.null digest then [] else ["digest" .= digest])

invalid :: Text -> IO a
invalid why = throwIO (PublicError "contract_invalid" why Nothing)
valid :: DeclarationIdentity -> IO ()
valid (DeclarationIdentity path digest) = do
  when (T.null path) (invalid "expected a nonempty declaration path")
  unless (T.null digest || T.length digest == 64 && T.all (\c -> c >= '0' && c <= '9' || c >= 'a' && c <= 'f') digest) (invalid "expected lowercase SHA-256 digest")
readIdentity :: Value -> IO DeclarationIdentity
readIdentity (Object members) = do
  unless (all (`elem` ["path","digest"]) (KM.keys members)) (invalid "unknown identity member")
  path <- case KM.lookup "path" members of Just (String p) -> pure p; _ -> invalid "expected path"
  digest <- case KM.lookup "digest" members of Nothing -> pure ""; Just (String d) | not (T.null d) -> pure d; _ -> invalid "invalid digest"
  let identity = DeclarationIdentity path digest
  valid identity
  pure identity
readIdentity _ = invalid "expected identity object"
compareIdentity :: DeclarationIdentity -> DeclarationIdentity -> IO ()
compareIdentity expected actual = when (declarationPath expected /= declarationPath actual || not (T.null (declarationDigest expected)) && not (T.null (declarationDigest actual)) && declarationDigest expected /= declarationDigest actual) $
  throwIO (PublicError "contract_mismatch" ("the declaration identity for " <> declarationPath expected <> " differs") Nothing)
identityHandler :: DeclarationIdentity -> IO Handler
identityHandler expected = do
  valid expected
  pure $ \_ _ value -> do
    actual <- readIdentity value
    compareIdentity expected actual
    pure (toJSON expected)
checkIdentity :: Peer -> CallContext -> DeclarationIdentity -> IO ()
checkIdentity peer ctx expected = do
  valid expected
  (call peer ctx "identity.check" (toJSON expected) >>= readIdentity >>= compareIdentity expected) `catch` \(err :: PublicError) ->
    unless (errorCode err == "method_not_found") (throwIO err)
